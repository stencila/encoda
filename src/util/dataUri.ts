/**
 * @module util/dataUri
 */

import { getLogger, LogEvent } from '@stencila/logga'
import fs from 'fs-extra'
import mime from 'mime'
import path from 'path'
import tempy from 'tempy'

const log = getLogger('encoda:util')

const DATA_URI_REGEX = /^data:([\w/+]+);(charset=[\w-]+|base64).*,(.*)/

/**
 * Test that a string is a Data URI
 *
 * @param str A string to test
 */
export function match(str: string): boolean {
  return DATA_URI_REGEX.test(str)
}

/**
 * Write a Data URI to file
 *
 * @param dataUri The data URI to write.
 * @param filePath The path of the file to write to (defaults to a temporary file)
 */
export async function toFile(
  dataUri: string,
  filePath?: string
): Promise<{ mediaType: string; filePath: string }> {
  const match = DATA_URI_REGEX.exec(dataUri)
  if (match === null) return { mediaType: '', filePath: '' }

  const mediaType = match[1]
  if (filePath === undefined) {
    const extension = mime.getExtension(mediaType) ?? 'png'
    filePath = tempy.file({ extension })
  }

  const data = dataUri.split(';base64,').pop()
  await fs.ensureDir(path.dirname(filePath))
  await fs.writeFile(filePath, data, 'base64')

  return { mediaType, filePath }
}

/**
 * Create a Data URI from file.
 *
 * @param filePath The file path to read.
 * @param mediaType The media type of the file (inferred from file extension if not provided)
 */
export async function fromFile(
  filePath: string,
  mediaType?: string
): Promise<{ mediaType: string; dataUri: string }> {
  if (mediaType === undefined) mediaType = mime.getType(filePath) ?? 'image/png'

  let data = ''
  try {
    data = await fs.readFile(filePath, 'base64')
  } catch (err) {
    const error = err as LogEvent & { code: string }
    if (error.code === 'ENOENT') {
      log.warn(`Image file does not exist, ignoring: ${filePath}`)
    } else {
      log.error(error)
    }
  }
  const dataUri = `data:${mediaType};base64,${data}`

  return { mediaType, dataUri }
}

/**
 * Create a Data URI from a JavaScript object.
 *
 * JSON stringifies and then base64 encodes the object.
 *
 * @param object The object to encode
 * @param mediaType The media type to encode into the Data URI
 */
export function fromJS(object: object, mediaType = 'application/json'): string {
  const json = JSON.stringify(object)
  const data = Buffer.from(json).toString('base64')
  return `data:${mediaType};base64,${data}`
}
