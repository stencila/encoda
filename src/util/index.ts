export const isDefined = <T>(n: T): n is NonNullable<typeof n> => n != null

export const reduceNonNullable =
  <A, B>(callback: (sourceNode: A) => B) =>
  (convertedNodes: NonNullable<B>[] = [], sourceNode?: A): NonNullable<B>[] => {
    const decodedNode =
      sourceNode !== undefined ? callback(sourceNode) : undefined
    return isDefined(decodedNode)
      ? [...convertedNodes, decodedNode]
      : convertedNodes
  }

/**
 * Removes all properties that are undefined from a given object.
 * Note that `null` values are considered explicit falsy values, and are not removed.
 */
export const compactObj = <O extends object>(o: O): O => {
  return Object.entries(o).reduce(
    (compacted: O, [k, v]) => {
      return v !== undefined ? { ...compacted, [k]: v } : compacted
    },
    {} as O, // eslint-disable-line
  )
}
