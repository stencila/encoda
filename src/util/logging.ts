import * as logga from '@stencila/logga'
import schema from '@stencila/schema'
import { deleteAt } from 'fp-ts/lib/Record'

const log = logga.getLogger('encoda')
export default log

/**
 * Configure logger so that it:
 *
 * - only shows DEBUG entries if `debug=true`
 * - does not show the same message twice within 5s
 *
 * @param debug In debug mode?
 */
export const configure = (debug = false, exitOnError = true): void => {
  logga.replaceHandlers((data: logga.LogData): void => {
    logga.defaultHandler(data, {
      maxLevel: debug ? logga.LogLevel.debug : logga.LogLevel.info,
      showStack: debug,
      exitOnError,
      throttle: {
        // Do not repeat the same message within 5s
        signature: `${data.tag}${data.level}${data.message}`,
        duration: 5000,
      },
    })
  })
}

type CodecOperation = 'decode' | 'encode'

/**
 * Log a warning about the loss of data
 *
 * ```ts
 * logWarn('html', 'decode', `Unhandled element type ${elem.name}`)
 * ```
 *
 * @param codec The codec emitting the warning
 * @param op The operation emitting the warning
 * @param message The message to emit
 */
export const logWarnLoss = (
  codec: string,
  op: CodecOperation,
  message: string,
): void => {
  log.warn(`${codec}:${op} ${message}`)
}

/**
 * Log a warning if there is any loss of data
 *
 * Use this function to warn the user of any loss of
 * data during encoding or decoding e.g.
 *
 * ```ts
 * const {used, ...lost} = node
 * logWarnLossIfAny('html', 'encode', node, lost)
 * ```
 *
 * @param codec The codec emitting the warning
 * @param op The operation emitting the warning
 * @param node The node from which data may be lost
 * @param lost An object with data that will be lost (if any)
 */
export const logWarnLossIfAny = (
  codec: string,
  op: CodecOperation,
  node: schema.Node,
  lost: object,
): void => {
  const rest = deleteAt('type')(lost)
  const properties = Object.keys(rest)
  if (properties.length > 0) {
    log.warn(
      `${codec}:${op} Properties of \`${schema.nodeType(
        node,
      )}\` not supported: ${properties.map((prop) => `\`${prop}\``).join(', ')}`,
    )
  }
}

/**
 * Log an error regarding unhandled / unexpected node type
 *
 * @param codec The codec emitting the warning
 * @param op The operation emitting the warning
 * @param types The type/s expected
 * @param node The node from which data may be lost
 */
export const logErrorNodeType = (
  codec: string,
  op: CodecOperation,
  types: string,
  node: schema.Node,
): void => {
  log.error(
    `${codec}:${op} Expecting node of type ${types} but got node of type ${schema.nodeType(
      node,
    )}`,
  )
}
