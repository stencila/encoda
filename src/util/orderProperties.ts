/**
 * @module util/orderProperties
 */

import schema from '@stencila/schema'
import { transformSync } from './transform'

/**
 * Transform an `Entity` node (and all it's descendant entities)
 * by ordering it's properties.
 *
 * The order is designed to be deterministic (i.e. not dependent upon
 * the order properties were defined in code) and human readable
 * (e.g. `type` first, often long `content` last).
 */
export default function orderProperties(node: schema.Node): schema.Node {
  return transformSync(node, (node: schema.Node): schema.Node => {
    if (!schema.isEntity(node)) return node
    const { type, id, meta, content, ...rest } = node as any
    return {
      // The node type and id are always first
      type,
      ...(id === undefined ? {} : { id }),
      // Other attributes in alphabetical order
      ...Object.keys(rest)
        .sort()
        .reduce((prev, curr) => ({ ...prev, ...{ [curr]: rest[curr] } }), {}),
      // Meta data can be long
      ...(meta === undefined ? {} : { meta }),
      // Content is often long and heavily nested so always last
      ...(content === undefined ? {} : { content }),
    }
  })
}
