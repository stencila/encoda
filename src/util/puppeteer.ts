/**
 * @module util/puppeteer
 */

import { getLogger } from '@stencila/logga'
import AsyncLock from 'async-lock'
import isDocker from 'is-docker'
import path from 'path'
import { Browser, Page } from 'puppeteer'
import isPackaged from './app/isPackaged'

export { Page, Browser }

// For unknown reasons it seems necessary to import like this
// eslint-disable-next-line @typescript-eslint/no-var-requires
const puppeteer = require('puppeteer')

const log = getLogger('encoda:puppeteer')

/**
 * The following code is necessary to ensure the Chromium binary can be correctly
 * found when bundled as a binary using [`pkg`](https://github.com/zeit/pkg).
 * See: [`pkg-puppeteer`](https://github.com/rocklau/pkg-puppeteer)
 */

// Adapts the regex path to work on both Windows and *Nix platforms
const pathRegex =
  process.platform === 'win32'
    ? /^.*?\\node_modules\\puppeteer\\\.local-chromium/
    : /^.*?\/node_modules\/puppeteer\/\.local-chromium/

export const executablePath = isPackaged
  ? puppeteer
      .executablePath()
      .replace(
        pathRegex,
        path.join(
          path.dirname(process.execPath),
          'node_modules',
          'puppeteer',
          '.local-chromium'
        )
      )
  : puppeteer.executablePath()

/**
 * Module global Puppeteer instance
 * and mutex lock to prevent conflicts between
 * concurrent requests to `startup` or `shutdown`
 */
let browser: Browser | undefined
const lock = new AsyncLock()

/**
 * Startup the browser if it isn't already.
 *
 * This needs to use a mutex lock to ensure that multiple
 * async calls to startup() don't race to create the
 * singleton browser instance.
 */
export async function startup(): Promise<Browser> {
  return lock.acquire('browser', async (): Promise<Browser> => {
    if (browser === undefined) {
      log.debug('Launching new browser')
      browser = await puppeteer.launch({
        pipe: true,
        executablePath,
        args: isDocker()
          ? [
              // Turn off Chrome's sandbox. The alternative is to make the container privileged
              // by adding `--cap-add=SYS_ADMIN` to `docker run`. But in the context's that Encoda normally
              // runs within a container, this is even more undesirable.
              // See the following for discussion https://github.com/docker/for-linux/issues/496#issuecomment-441149510
              '--no-sandbox',
              '--disable-setuid-sandbox',
              // Use /tmp instead of /dev/shm to avoid issues like: https://dev.azure.com/stencila/stencila/_build/results?buildId=205&view=logs&j=b17395f6-68a3-5682-0476-d3f6f1043109&t=e59dc482-4022-5828-e063-e9c9e022e048&l=440
              // See https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#tips
              '--disable-dev-shm-usage',
            ]
          : [],
      })
      if (browser === undefined) {
        throw new Error('Unable to launch browser instance')
      }
      log.debug(`Browser launched. pid: ${browser.process()?.pid}`)
    }
    return browser
  })
}

/**
 * Create a new page
 */
export async function page(): Promise<Page> {
  const browser = await startup()
  return browser.newPage()
}

/**
 * Close the browser.
 */
export async function shutdown(): Promise<void> {
  await lock.acquire('browser', async (): Promise<void> => {
    if (browser !== undefined) {
      log.debug(`Closing browser. pid: ${browser.process()?.pid}`)
      await browser.close()
      log.debug('Browser closed')
      browser = undefined
    }
  })
}

// Always shutdown before exiting the Node process
// We use `beforeExit` because async operations are not supported
// by `exit`.
// See https://nodejs.org/api/process.html#process_event_beforeexit
process.on('beforeExit', () => {
  shutdown().catch((error) => {
    throw error
  })
})
