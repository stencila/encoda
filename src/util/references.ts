import { CreativeWork, isA, Organization, Person } from '@stencila/schema'
import { array as A } from 'fp-ts'

/**
 * Create string for use in "numeric" in-text citations e.g. `[1]`.
 *
 * Will return "-1" if there are no references and "0" if the work
 * is not able to be found in the references
 */
export const encodeCiteNumeric = (
  work: CreativeWork | string,
  references: CreativeWork['references'],
): string => {
  return `${(references?.indexOf(work) ?? -2) + 1}`
}

/**
 * Create string for the authors part of in-text citations.
 * e.g. `Smith et al.` in `Smith et al. (1990)`.
 */
export const encodeCiteAuthors = (work: CreativeWork): string | undefined => {
  const { authors = [] } = work

  if (A.isEmpty(authors)) return undefined

  let citeText = ''
  const firstAuthorName = getName(authors[0])
  if (firstAuthorName.length > 0) {
    citeText += firstAuthorName
    if (authors.length === 2) {
      const secondAuthorName = getName(authors[1])
      if (secondAuthorName.length > 0) citeText += ' and ' + secondAuthorName
    } else if (authors.length > 2) {
      citeText += ' et al.'
    }
  }
  return citeText
}

/**
 * Create string for the year part on in-text citations.
 * e.g. the `1990` in `Smith et al. (1990)`.
 */
export const encodeCiteYear = (work: CreativeWork): string | undefined => {
  const { datePublished } = work

  if (datePublished === undefined) return undefined

  const date =
    typeof datePublished === 'string' ? datePublished : datePublished.value
  return date.split('-')[0]
}

/**
 * Create string for "author-year" type citations within
 * a citation group e.g. `Smith et al., 1990`.
 */
export const encodeCiteAuthorsYear = (work: CreativeWork): string => {
  const authors = encodeCiteAuthors(work) ?? ''
  const year = encodeCiteYear(work)
  return year !== undefined ? `${authors}, ${year}` : authors
}

/**
 * Get the name of a person or organization to use in an in-text citation.
 */
function getName(author: Person | Organization): string {
  return isA('Person', author) &&
    author.familyNames &&
    author.familyNames.length > 0
    ? author.familyNames.join(' ').trim()
    : (author.name ?? '').trim()
}
