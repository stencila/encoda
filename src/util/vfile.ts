/**
 * @module util/vfile
 */

/**
 * Hello contributor üëã! If you are working on this file, please
 * endeavor to remove the need for the following `eslint-disable` line üôè.
 * Remove the line and run `npx eslint path/to/this/file.ts` to
 * see which code needs some linting ‚ù§Ô∏è.
 * See https://github.com/stencila/encoda/issues/199 for suggestions
 * on how to refactor code to avoid non-strict boolean expressions.
 */
/* eslint-disable @typescript-eslint/strict-boolean-expressions */

import fs from 'fs-extra'
import path from 'path'
import getStdin from 'get-stdin'
// @ts-ignore
import toVFile from 'to-vfile'
import vfile, { VFileContents } from 'vfile'

/**
 * The virtual file type
 */
export type VFile = vfile.VFile

/**
 * Create a virtual file
 *
 * @param contents Raw string contents of `VFile`, or file path
 * @param options Options to `vfile` see https://github.com/vfile/vfile#vfileoptions
 */
export function create(
  contents?: string,
  options: vfile.VFileOptions = {},
): VFile {
  if (contents) {
    if (isPath(contents)) options = { ...options, path: contents }
    else options = { ...options, contents }
  }
  return vfile(options)
}

/**
 * Load a string into a virtual file
 *
 * @param contents Contents to load
 */
export function load(
  contents: VFileContents,
  options: vfile.VFileOptions = {},
): VFile {
  return vfile({ ...options, contents })
}

/**
 * Dump a string from the contents of a virtual files
 *
 * If the file has no `contents` but does have a file
 * `path`, then `read` the virtual file.
 *
 * @param file The virtual file to dump
 */
// Not supported by eslint, worksaround doesn't work
// See https://github.com/benmosher/eslint-plugin-import/issues/1357
// eslint-disable-next-line
export async function dump(vfile: VFile, mode?: 'string'): Promise<string>
// eslint-disable-next-line
export async function dump(vfile: VFile, mode?: 'buffer'): Promise<Buffer>
// eslint-disable-next-line
export async function dump(
  vfile: VFile,
  mode: 'string' | 'buffer' = 'string',
): Promise<string | Buffer> {
  if (vfile.contents) {
    if (mode === 'string') return vfile.toString()
    else
      return Buffer.isBuffer(vfile.contents)
        ? vfile.contents
        : Buffer.from(vfile.contents)
  }
  if (vfile.path) {
    const readFile = await toVFile.read(vfile.path)
    return mode === 'string' ? readFile.toString() : readFile.contents
  }
  return mode === 'string' ? '' : Buffer.alloc(0)
}

/**
 * Read a file path into a virtual file
 *
 * @param path The content, or file system path to read from. If `-`, read from stdin.
 */
export async function read(content: string): Promise<VFile> {
  if (!content) throw new Error('Argument `path` is required')

  if (content === '-') {
    return load(await getStdin())
  } else if (isPath(content)) {
    return toVFile.read(content)
  } else {
    return load(content)
  }
}

/**
 * Write a virtual file to the file system or standard output.
 *
 * Will never write an empty file i.e. will only write if the
 * `vfile` has `contents`.
 *
 * @param vfile Virtual file to write
 * @param filePath The file system path to write to, or `-` to write to stdout.
 */
export async function write(vfile: VFile, filePath: string): Promise<void> {
  if (!filePath) throw new Error('Argument `filePath` is required')

  if (filePath === '-') {
    console.log(await dump(vfile))
  } else if (filePath && vfile.contents) {
    await fs.ensureDir(path.dirname(filePath))
    vfile.path = filePath
    await toVFile.write(vfile)
  }
}

/**
 * Is a string a file system path?
 *
 * Several functions in this module allow for content to be passed as a
 * string of raw content (e.g. `This is some Markdown`) or as a file system
 * path. This function, assesses whether a file is a file system path or not.
 *
 * If the string starts with `/`, `./`, `../` (or Windows compatible backslash
 * equivalents as well as drive letter prefixed strings e.g. `C:\`),
 * or has a 1-5 character extension,
 * then it is assumed to be a path but the presence of the path is not checked.
 * The reason for not checking presence
 * here is because "if the content looks like a path then the user probably meant
 * it to be a path". That is, if a user tries to convert the string "./file.md" then,
 * if that file doesn't exist, they probably want the Markdown converter to give me an error
 * message. They probably don't want the `match` function to try and find some other converter
 * that acts on a plain strings.
 *
 * For all other strings, this function does check for presence, returning `true`
 * if the file exists.
 *
 * @param content The string to assess.
 */
export function isPath(content: string): boolean {
  if (/^(\/)|(\\)|([A-Z]:\\)|(\.(\/|\\))|(\.\.(\/|\\))/.test(content)) {
    return true
  }
  if (fs.existsSync(content)) return true
  return false
}
