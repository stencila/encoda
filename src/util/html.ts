/**
 * Utility functions for processing HTML.
 *
 * This module provides some simple convenience functions for
 * processing HTML documents with a similar API to the
 * `util/xml` module.
 *
 * @module util/html
 */

import { getTheme, styleEntry, scriptEntry, ThemaAssets } from '@stencila/thema'
import fs from 'fs'
import jsdom from 'jsdom'
import path from 'path'
import { toFile } from './uri'
import { isPath } from './vfile'
import log from './logging'

const JSDOM = new jsdom.JSDOM()

/**
 * Parse HTML into a `HTMLDocument`
 */
export const load = (html: string): HTMLDocument => {
  const dom = new jsdom.JSDOM(html)
  return dom.window.document
}

/**
 * Create a `HTMLElement`.
 */
export const elem = (
  tagName: string,
  attrs?: { [key: string]: string | number | boolean },
  ...children: (string | HTMLElement)[]
): HTMLElement => {
  const elem = JSDOM.window.document.createElement(tagName)
  if (attrs !== undefined)
    for (const [key, value] of Object.entries(attrs))
      elem.setAttribute(key, value.toString())
  for (const child of children) {
    if (typeof child === 'string')
      elem.appendChild(JSDOM.window.document.createTextNode(child))
    else elem.appendChild(child)
  }
  return elem
}

/**
 * Get an element's text content.
 */
export const text = (elem: Node | null): string | null => {
  if (elem === null) return null
  return elem.textContent
}

/**
 * Get the first descendent element that matches the selector
 */
export const first = (
  elem: HTMLElement | null,
  selector = '*',
): HTMLElement | null => {
  if (elem === null) return null
  return elem.querySelector(selector)
}

/**
 * Get all descendent elements that match the selector
 */
export const all = (
  elem: HTMLElement | null,
  selector = '*',
): HTMLElement[] => {
  if (elem === null) return []
  return Array.from(elem.querySelectorAll(selector))
}

/**
 * Get all descendent nodes that have a given name e.g. `figure`, `#text`, `#comment`
 */
export const allName = (elem: HTMLElement | null, name: string): Node[] => {
  if (elem === null) return []
  return Array.from(elem.childNodes).reduce((prev: Node[], node) => {
    return [
      ...prev,
      ...(node.nodeName === name ? [node] : []),
      ...(node.nodeType === node.ELEMENT_NODE
        ? allName(node as HTMLElement, name)
        : []),
    ]
  }, [])
}

const themeNotFound = (themeName: string): ThemaAssets => {
  log.warn(`Theme assets could not be found for theme: '${themeName}'`)

  return {
    styles: [],
    scripts: [],
  }
}

/**
 * Given an array of filepaths or URLs, returns the file contents as strings.
 *
 * @async
 * @param {string[]} assets - Filepaths or URLs to read
 * @return {Promise<string[]>} - Promise resolving to the file contents
 */
const fetchAssets = async (assets: string[]): Promise<string[]> =>
  Promise.all(
    assets.map((asset) =>
      toFile(asset).then((file) => fs.readFileSync(file.filePath).toString()),
    ),
  )

/**
 * Given a theme name, path, or URL, attempt to resolve it into theme Stylesheet and JavaScript entry files
 * (`style.css` and `index.js`).
 *
 * @param {string} theme - Can be a Thema theme name, or a URL/filepath to a directory following Thema theme conventional structure
 * @return {ThemaAssets} - An object containing all the themeâ€˜s CSS and JavaScript files as an array.
 */
export const getThemeAssets = async (
  theme: string,
  isBundle = false,
): Promise<ThemaAssets> => {
  let resolvedTheme = getTheme(theme, !isBundle)

  // If theme could be resolved, it is a Thema theme and no other processing needs to be done
  if (resolvedTheme !== undefined) {
    return resolvedTheme
  }

  /**
   * If the given `theme` string does not end with a file extensions,
   * use the last part of the path as the directory name
   */
  const themeDir = (dir: string): string =>
    path.extname(dir) === '' ? path.basename(dir) : ''

  const getThemePath = (dir: string, file: string): string =>
    path.join(path.dirname(dir), themeDir(dir), file)

  // If theme is a URL, use it as a directory to look for theme assets based on Thema naming conventions
  // looking for a `styles.css` and `index.js` files.
  if (theme.includes('://')) {
    // Process the theme URL to clean trailing slashes or file names
    const cleanUrl = theme.endsWith('/')
      ? theme.slice(0, theme.length - 1)
      : theme
    const parts = cleanUrl.split('/')
    const endsInFile = /\.\w+$/.test(parts[parts.length - 1])
    // Account for receiving top level URL such as 'http://example.com', and not consider '.com' as a file extension
    const urlParts =
      parts.length > 3 && endsInFile ? parts.slice(0, parts.length - 1) : parts
    const url = urlParts.join('/')

    resolvedTheme = {
      styles: [`${url}/${styleEntry}`],
      scripts: [`${url}/${scriptEntry}`],
    }
  }

  // If theme is a filepath, and contains stylesheet and JavaScript entry files, it is a valid Thema theme
  if (
    isPath(theme) &&
    fs.existsSync(path.join(theme, styleEntry)) &&
    fs.existsSync(path.join(theme, scriptEntry))
  ) {
    resolvedTheme = {
      styles: [getThemePath(theme, styleEntry)],
      scripts: [getThemePath(theme, scriptEntry)],
    }
  }

  // If a theme could not be resolved, return an empty theme and log a warning message
  if (resolvedTheme === undefined) {
    return themeNotFound(theme)
  }

  // If a theme is to be bundled, read the contents of the theme files
  const styles = isBundle
    ? await fetchAssets(resolvedTheme.styles)
    : resolvedTheme.styles
  const scripts = isBundle
    ? await fetchAssets(resolvedTheme.scripts)
    : resolvedTheme.scripts

  return {
    styles,
    scripts,
  }
}
