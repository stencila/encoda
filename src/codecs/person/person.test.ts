import schema from '@stencila/schema'
import { dump } from '../../util/vfile'
import { nockRecord } from '../../__tests__/helpers'
import { PersonCodec } from './'

const { sniff, decode, encode } = new PersonCodec()

test('sniff', async () => {
  expect(await sniff('Joe Jones')).toBe(true)
  expect(await sniff('J Jones')).toBe(true)
  expect(await sniff('Joe James Jones')).toBe(true)
  expect(await sniff('Joe J. Jones')).toBe(true)
  expect(await sniff('Joe J. Jones <joe@example.com>')).toBe(true)
  expect(await sniff('Joe J. Jones <joe-jones@example.com>')).toBe(true)

  expect(await sniff('Joe')).toBe(false)
  expect(await sniff('joe Jones')).toBe(false)
  expect(await sniff('Lorem ipsum')).toBe(false)
})

const joe = {
  content: 'Joe Jones',
  node: {
    type: 'Person',
    givenNames: ['Joe'],
    familyNames: ['Jones'],
  },
}

const jane = {
  content: 'Jane Jennifer Jones <jane@example.com>',
  node: {
    type: 'Person',
    givenNames: ['Jane', 'Jennifer'],
    familyNames: ['Jones'],
    emails: ['jane@example.com'],
  },
}

const jill = {
  content: 'Dr Jill J. Jones PhD <jill@example.com> (http://example.com/jill)',
  node: {
    type: 'Person',
    givenNames: ['Jill', 'J.'],
    familyNames: ['Jones'],
    emails: ['jill@example.com'],
    url: 'http://example.com/jill',
    honorificPrefix: 'Dr',
    honorificSuffix: 'PhD',
  },
}

describe('decode', () => {
  it('handles various name parts', async () => {
    let person = schema.person()

    person.familyNames = ['Jones']
    expect(await decode('Jones')).toEqual(person)

    person.givenNames = ['Jane', 'Jill']
    expect(await decode('Jane Jill Jones')).toEqual(person)

    person.honorificPrefix = 'Dr'
    expect(await decode('Dr Jane Jill Jones')).toEqual(person)

    person.honorificSuffix = 'PhD'
    expect(await decode('Dr Jane Jill Jones PhD')).toEqual(person)

    person.emails = ['jane@example.com']
    expect(await decode('Dr Jane Jill Jones PhD <jane@example.com>')).toEqual(
      person
    )

    person.url = 'http://example.com/jane'
    expect(
      await decode(
        'Dr Jane Jill Jones PhD <jane@example.com> (http://example.com/jane)'
      )
    ).toEqual(person)
  })

  it('decodes examples', async () => {
    expect(await decode(joe.content)).toEqual(joe.node)
    expect(await decode(jane.content)).toEqual(jane.node)
    expect(await decode(jill.content)).toEqual(jill.node)
  })

  it('decodes an orcid to a Person', async () => {
    const done = await nockRecord('nock-record-orcid.json')

    expect(
      await decode('https://orcid.org/0000-0002-1825-0097')
    ).toMatchSnapshot()

    done()
  })

  it('returns an empty person if name can not be parsed', async () => {
    expect(await decode('')).toEqual(schema.person())
    expect(await decode('#@&%')).toEqual(schema.person())
  })
})

test('encode', async () => {
  expect(await dump(await encode(joe.node))).toEqual(joe.content)
  expect(await dump(await encode(jane.node))).toEqual(jane.content)
  expect(await dump(await encode(jill.node))).toEqual(jill.content)
})
