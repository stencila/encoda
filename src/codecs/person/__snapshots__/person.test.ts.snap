// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode decodes an orcid to a Person 1`] = `
{
  "affiliations": [
    {
      "alternateNames": "Psychoceramics",
      "identifiers": {
        "propertyID": "RINGGOLD",
        "type": "PropertyValue",
        "value": "5468",
      },
      "name": "Wesleyan University",
      "type": "Organization",
    },
    {
      "alternateNames": "Psychoceramics",
      "identifiers": {
        "propertyID": "RINGGOLD",
        "type": "PropertyValue",
        "value": "6752",
      },
      "name": "Brown University",
      "type": "Organization",
    },
  ],
  "alternateNames": [
    "Josiah Stinkney Carberry",
    "J. Carberry",
    "J. S. Carberry",
  ],
  "familyNames": "Carberry",
  "givenNames": "Josiah",
  "id": "https://orcid.org/0000-0002-1825-0097",
  "identifiers": {
    "propertyID": "Scopus Author ID",
    "type": "PropertyValue",
    "value": "7007156898",
  },
  "schema:mainEntityOfPage": {
    "id": "https://orcid.org/0000-0002-1825-0097",
  },
  "type": "Person",
  "url": [
    {
      "id": "http://library.brown.edu/about/hay/carberry.php",
    },
    {
      "id": "http://en.wikipedia.org/wiki/Josiah_Carberry",
    },
  ],
}
`;
